### PR TITLE
Simplify SHM-allocator

### PR DESCRIPTION
**Pull request by gbalykov**
This pull request simplifies SHM-allocator

Related issue: #12765 

On application, which is referenced in #10380 (comment), the following memory consumption changes occur (based on 23a09c98aa2cace8616f360ee393837b1701b7f8):

Rss: 44936 -> 44112 (1.80% improvement)
Private_Dirty: 20936 -> 20160 (3.71% improvement)
Private_Clean: 472 -> 472
Shared_Clean: 23408 -> 23360

@Dmitri-Botcharnikov @ruben-ayrapetyan

### CODE REVIEW

**Comment by gbalykov** [comment]
@jkotas

Could you, please, take a look?

**Comment by adityamandaleeka** [comment]
Thanks for doing this @gbalykov. It looks like most of the changes are just code deletion, and a lot (or all) of this code is never called in the first place. Am I right in assuming that those memory consumption improvements you posted are primarily due to reduced code size?

Are those before/after numbers for a Release build?

**Comment by gbalykov** [comment]
@adityamandaleeka 

Numbers are for Release build for Tizen ARM.

Code size changes are related to reduction of Shared_Clean by 48 kb. Private_Dirty changes occur due to removal of two segments of 256 kb size, which were previously allocated by SHM-allocator. Additional memory reduction at approximately 256 kb happens due to unknown reason.

**Comment by adityamandaleeka** [comment]
Thanks @gbalykov. This change looks fine to me.

**Comment by adityamandaleeka** [comment]
Actually, I would prefer if this change went one step further and removed things like SHMalloc entirely instead of just changing the implementation to use malloc. When I left my previous comment, I thought this is something that could be done in the future, but I see that it's only used in a few places, so it is better to extract it properly now.

File: src/pal/src/include/pal/shmemory.h [hunk 1]
```diff
@@ -17,55 +17,10 @@ How to use :
 
 The SHMalloc function can be used to allocate memory in the shared memory area.
```
Comment (adityamandaleeka) [inline src/pal/src/include/pal/shmemory.h]: This needs to be updated.

File: src/pal/src/shmemory/shmemory.cpp [hunk 2]
```diff
@@ -1466,23 +467,9 @@ SHMPTR SHMStrDup( LPCSTR string )
...
-                // there's a problem in either the macro, or the underlying
-                // call to SHMPtrToPtr. In case the impossible happens,
-                // though, free the memory and return NULL rather than
-                // returning uninitialized memory.
-                SHMfree( retVal );
-                retVal = NULL;
-            }
+            LPVOID ptr = SHMPTR_TO_PTR(retVal);
+            _ASSERTE(ptr != NULL);
+            memcpy(ptr, string, length);
```
Comment (adityamandaleeka) [inline src/pal/src/shmemory/shmemory.cpp]: `SHMStrDup` and `SHMWStrDup` should also just be removed entirely.

**Comment by gbalykov** [comment]
@dotnet-bot help

**Comment by gbalykov** [comment]
@dotnet-bot test Windows_NT arm Cross Release Build

**Comment by gbalykov** [comment]
@adityamandaleeka 

I have removed SHMalloc, SHMfree, SHMStrDup, SHMWStrDup and also made SHMPTR a LPVOID, so now no cast to/from SHMPTR is needed.

File: src/pal/src/include/pal/shmemory.h (hunk 1 above)
Comment (gbalykov) [inline src/pal/src/include/pal/shmemory.h]: Fixed

File: src/pal/src/shmemory/shmemory.cpp (hunk 2 above)
Comment (gbalykov) [inline src/pal/src/shmemory/shmemory.cpp]: Fixed

File: src/pal/src/synchmgr/synchmanager.cpp [hunk 3]
```diff
@@ -4091,8 +4091,7 @@ namespace CorUnix
         pthread_condattr_t attrs;
         pthread_condattr_t *attrsPtr = nullptr;
 
-        m_shridWaitAwakened = RawSharedObjectAlloc(sizeof(DWORD),
-                                                   DefaultSharedPool);
+        m_shridWaitAwakened = malloc(sizeof(DWORD));
         if (NULLSharedID == m_shridWaitAwakened)
```
Comment (adityamandaleeka) [inline src/pal/src/synchmgr/synchmanager.cpp]: This can just check for NULL.

Comment (adityamandaleeka) [inline src/pal/src/synchmgr/synchmanager.cpp]: (and in other places too)

**Comment by adityamandaleeka** [comment]
This looks fine to me, but I'll wait for a second review before merging. @janvorli Can you take a look when you have some time?

**Comment by adityamandaleeka** [comment]
Thanks for making those changes @gbalykov

**Comment by gbalykov** [comment]
@dotnet-bot test Tizen armel Cross Release Build

File: src/pal/src/synchmgr/synchmanager.cpp (hunk 3 above)
Comment (gbalykov) [inline src/pal/src/synchmgr/synchmanager.cpp]: Fixed

**Pull request merged**